### ISSUE TITLE
Add actions in pay_action() method

### ISSUE THREAD

**Issue by Dev49net** [woothemes/woocommerce#10276]
There are no actions or filters in **_pay_action()_** in **includes/class-wc-form-handler.php**, which makes it impossible to tweak using plugins. How about adding two actions: **woocommerce_before_pay_action** and **woocommerce_after_pay_action**, passing _$order_ and possibly _$order_key_ as arguments?

**Comment by mikejolley** [comment]
The actions in your PR would run even if the order key was wrong/invalid.

If you re-raise it after resolving the above, let me know the use case for such a chance. "Tweaking" the pay action doesn't seem desirable and I fear could just make debugging of payments ten times harder if something silly is hooked in.

Thanks

**Comment by Dev49net** [comment]
Hi Mike, thanks a lot for your input. 

I usually like very early hooks, so that developers can act as early as possible, but in this case you are right - it is better to do this after the verifying conditional statement. I will submit another PR when I move these actions inside the conditional statement.

I am a developer of multi currency plugin for WooCommerce and I have encountered a problem I cannot solve without these actions. Basically - I am creating an order from the backend (WooCommerce > Orders > Add order). I am setting the order currency to currency other than base currency (let's assume the base currency is USD and I am creating a specific custom order for my British customer in GBP). I create the order and run the action Email customer invoice. Customer gets invoice to his email and a payment link - all data is in GBP. And here is the problem - the payment will have the same values I entered, but it will not be in GBP, but USD instead (even though the payment form summary will show GBP; PayPal/etc. will get currency as USD, because payment processors get currency from `get_woocommerce_currency()`). 

So my solution to this is to use these actions I am proposing here. Right before payment processor gets the data (`woocommerce_before_pay_action` action) I will hook a filter into `woocommerce_currency`, change the currency to the one I take from `$order->get_order_currency()`, so that when PayPal/etc. get currency by using `get_woocommerce_currency()` it will get the one filtered by me. Then thanks to the next action (`woocommerce_after_pay_action`) I will unhook my `woocommerce_currency` filter so that in does not affect the whole store - just the data sent to payment gateway.

So basically what I need is to filter `woocommerce_currency`, but only for the data that is sent to payment gateway.

Daniel

**Comment by daigo75** [comment]
@Dev49net I dealt with the same issue when I developed my Currency Switcher, in 2013. You can already determine if the customer is paying for an order without having to add extra filters (I would just have to remember how I did it).

**Comment by Dev49net** [comment]
Hi @daigo75, thanks for your input. Well, actually, I don't think it is possible without that action that I committed (will be included in WC 2.6). Please keep in mind that there are two ways to pay for an order - one is when user is placing an order and then normally going to checkout - which I did not have any trouble with. But it is something different when you create an order from the dashboard in custom currency and then send a payment link. Different procedure is used for that. I studied the whole process really carefully and I have come to conclusion that introducing additional actions is the only way to make it work.

**Comment by daigo75** [comment]
@Dev49net I know what you mean. I added support for orders created manually over two years ago, and I know that I covered all cases (normal checkout and payment link), ever since WC 2.0. As I wrote, I will just have to remember how I did it. :smile:

**Comment by Dev49net** [comment]
@daigo75 Cheers, if you find out how you did it, I would appreciate some guidance :-) Either way, I already have a solution based on the hook I proposed, so I am just waiting for 2.6 to introduce it.

**Comment by daigo75** [comment]
With the new actions it will definitely be easier (although I won't be able to rely on them, as I must guarantee backward compatibility). I will let you know if I manage to dig out the "trick" I used. I wrote it so long ago that I can't remember the approach I took (and it still works).
